ray-curator/ray_curator/stages/deduplication/fuzzy/generate_duplicate_ids.py

```diff
@@ -0,0 +1,142 @@
+        If None spilling is disabled.
+    enable_statistics
+        Whether the underlying rapidsmpf shuffler should collect shuffle statistics.
+    """
+
+    _name = "RemovalIDs"
+
+    def __init__(  # noqa: PLR0913
+        self,
+        duplicate_group_field: str = CURATOR_FUZZY_DUPLICATE_GROUP_FIELD,
+        document_id_field: str = CURATOR_DEDUP_ID_STR,
+        total_nparts: int | None = None,
+        output_path: str = "./",
+        read_kwargs: dict[str, Any] | None = None,
+        write_kwargs: dict[str, Any] | None = None,
+        rmm_pool_size: int | Literal["auto"] | None = "auto",
+        spill_memory_limit: int | Literal["auto"] | None = "auto",
+        enable_statistics: bool = False,
+    ):
+        self.duplicate_group_field = duplicate_group_field
+        self.document_id_field = document_id_field
+        self.output_fs = get_fs(
+            output_path, storage_options=read_kwargs.get("storage_options") if read_kwargs is not None else None
+        )
+        self.output_path = self.output_fs.sep.join([output_path, self.name])
+        self.write_kwargs = write_kwargs
+
+        super().__init__(
+            shuffle_on=[duplicate_group_field],
+            total_nparts=total_nparts,
+            output_path=self.output_path,
+            read_kwargs=read_kwargs,
+            write_kwargs=write_kwargs,
+            rmm_pool_size=rmm_pool_size,
+            spill_memory_limit=spill_memory_limit,
+            enable_statistics=enable_statistics,
+        )
+
+    def _get_removal_ids(self, df: "cudf.DataFrame") -> "cudf.DataFrame":
+        """
+        Get the removal ids for the given dataframe.
+        """
+        removal_ids = df[df[self.duplicate_group_field].duplicated(keep="first")][self.document_id_field]
+        removal_ids = removal_ids.sort_values(ignore_index=True)
+        return removal_ids.to_frame()
+
+    def process(self, task: FileGroupTask) -> FileGroupTask:
+        return super().process(task)
+
+    def ray_stage_spec(self) -> dict[str, Any]:
+        return super().ray_stage_spec()
+
+    def read_and_insert(self, task: FileGroupTask) -> FileGroupTask:
+        super().read_and_insert(task)
+        return task
+
+    def insert_finished(self) -> None:
+        super().insert_finished()
+
+    def extract_and_write(self) -> list[FileGroupTask]:
+        self._check_actor_obj()
+        write_kwargs = self.write_kwargs.copy()
+        write_kwargs["index"] = write_kwargs.get("index", False)
+
+        result_tasks = []
+        for partition_id, partition in self._actor_obj.extract():
+            shuffled_partition_df = pylibcudf_to_cudf_dataframe(partition, column_names=self.output_columns)
+            num_groups = shuffled_partition_df[self.duplicate_group_field].nunique()
+            removal_ids = self._get_removal_ids(shuffled_partition_df)
+
+            output_file = self.output_fs.sep.join([self.output_path, f"part.{partition_id}.parquet"])
```

Comment:
I'm unclear on the the details of a ShuffleStage, but is partition_id guaranteed to be unique across actors; i.e. is there a case where we have different processes overwriting the same files

Reply:
Yes `partition_id` is guaranteed to be unique among actors. If a shuffle has `N` output partitions the shuffler assigns `~N/K` partition_id's to each actor/process. That actors receives the specific partition post shuffle.